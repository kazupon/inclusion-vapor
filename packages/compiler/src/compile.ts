// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `@vue/compiler-vapor`
// Author: Evan you (https://github.com/yyx990803) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/vuejs/core-vapor
// Code url: https://github.com/vuejs/core-vapor/blob/6608bb31973d35973428cae4fbd62026db068365/packages/compiler-vapor/src/compile.ts

import { parse } from 'svelte/compiler'
import { ErrorCodes, createCompilerError, defaultOnError } from '@vue-vapor/compiler-dom'
import { generate } from '@vue-vapor/compiler-vapor'
import { extend, isString } from '@vue-vapor/shared'
import {
  transformChildren,
  transformComment,
  transformElement,
  transformText,
  transformVBind,
  transformVOn
} from './transforms'
import { transform } from './transform'
import { IRNodeTypes } from './ir'

import type {
  CompilerOptions as BaseCompilerOptions,
  VaporCodegenResult,
  RootIRNode as VaporRootIRNode
} from '@vue-vapor/compiler-vapor'
import type { SvelteAst, RootNode } from './ir'
import type { HackOptions, NodeTransform, DirectiveTransform } from './transforms'

// Svelte Code / Svelte AST -> IR (transform) -> JS (generate)
export function compile(
  source: string | SvelteAst,
  options: CompilerOptions = {} // TODO: maybe we need some svelte compiler options
): VaporCodegenResult {
  const onError = options.onError || defaultOnError
  const isModuleMode = options.mode === 'module'
  /* istanbul ignore if */
  if (__BROWSER__) {
    if (options.prefixIdentifiers === true) {
      onError(createCompilerError(ErrorCodes.X_PREFIX_ID_NOT_SUPPORTED))
    } else if (isModuleMode) {
      onError(createCompilerError(ErrorCodes.X_MODULE_MODE_NOT_SUPPORTED))
    }
  }

  const prefixIdentifiers = !__BROWSER__ && (options.prefixIdentifiers === true || isModuleMode)

  if (options.scopeId && !isModuleMode) {
    onError(createCompilerError(ErrorCodes.X_SCOPE_ID_NOT_SUPPORTED))
  }

  const resolvedOptions = extend({}, options, {
    prefixIdentifiers
  })

  // TODO: tweak options for svelte parser
  const svelteAst = isString(source) ? parse(source) : source
  // TODO: convert `instance` and `module` to the Vapor (Vue) runtime API
  // svelteAst.instance
  // svelteAst.module

  const ast: RootNode = {
    type: IRNodeTypes.ROOT,
    children: svelteAst.html.children || [],
    source: isString(source) ? source : '', // TODO
    components: [],
    directives: [],
    helpers: new Set(),
    temps: 0
  }

  const [nodeTransforms, directiveTransforms] = getBaseTransformPreset(prefixIdentifiers)

  const ir = transform(
    ast,
    extend({}, resolvedOptions, {
      nodeTransforms: [
        ...nodeTransforms,
        ...(options.nodeTransforms || []) // user transforms
      ],
      directiveTransforms: extend(
        {},
        directiveTransforms,
        options.directiveTransforms || {} // user transforms
      )
    })
  )

  return generate(ir as unknown as VaporRootIRNode, resolvedOptions)
}

export type CompilerOptions = HackOptions<BaseCompilerOptions>
export type TransformPreset = [NodeTransform[], Record<string, DirectiveTransform>]

export function getBaseTransformPreset(_prefixIdentifiers?: boolean): TransformPreset {
  return [
    [
      // transformVOnce,
      // transformVIf,
      // transformVFor,
      // transformSlotOutlet,
      // transformTemplateRef,
      transformText,
      transformElement,
      // transformVSlot,
      transformComment,
      transformChildren
    ],
    {
      bind: transformVBind,
      on: transformVOn
      // html: transformVHtml,
      // text: transformVText,
      // show: transformVShow,
      // model: transformVModel,
    }
  ]
}
