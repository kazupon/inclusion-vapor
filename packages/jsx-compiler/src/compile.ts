// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `unplugin-vue-jsx-vapor`
// Author: zhiyuanzmj (https://github.com/zhiyuanzmj) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/unplugin/unplugin-vue-jsx-vapor
// Code url: https://github.com/unplugin/unplugin-vue-jsx-vapor/tree/main/src/core/compiler

import { parse } from '@babel/parser'
import { ErrorCodes, createCompilerError, defaultOnError } from '@vue-vapor/compiler-dom'
import { generate } from '@vue-vapor/compiler-vapor'
import { extend, isString } from '@vue-vapor/shared'
import { IRNodeTypes } from './ir/index.ts'
import { transform } from './transform.ts'
import {
  transformBind,
  transformChildren,
  transformElement,
  transformOn,
  transformText
} from './transforms/index.ts'

import type {
  CompilerOptions as BaseCompilerOptions,
  VaporCodegenResult as BaseVaporCodegenResult,
  RootIRNode as VaporRootIRNode
} from '@vue-vapor/compiler-vapor'
import type { BabelProgram, JSXElement, JSXFragment, RootIRNode, RootNode } from './ir/index.ts'
import type { DirectiveTransform, HackOptions, NodeTransform } from './transforms/index.ts'

export interface VaporCodegenResult extends Omit<BaseVaporCodegenResult, 'ast'> {
  ast: RootIRNode
}

// code/AST -> IR (transform) -> JS (generate)
export function compile(
  source: string | BabelProgram,
  options: CompilerOptions = {}
): VaporCodegenResult {
  const onError = options.onError || defaultOnError
  const isModuleMode = options.mode === 'module'
  /* istanbul ignore if */
  if (__BROWSER__) {
    if (options.prefixIdentifiers === true) {
      onError(createCompilerError(ErrorCodes.X_PREFIX_ID_NOT_SUPPORTED))
    } else if (isModuleMode) {
      onError(createCompilerError(ErrorCodes.X_MODULE_MODE_NOT_SUPPORTED))
    }
  }

  const prefixIdentifiers = !__BROWSER__ && options.prefixIdentifiers === true

  if (options.scopeId && !isModuleMode) {
    onError(createCompilerError(ErrorCodes.X_SCOPE_ID_NOT_SUPPORTED))
  }

  const resolvedOptions = extend({}, options, {
    prefixIdentifiers,
    expressionPlugins: options.expressionPlugins || ['jsx']
  })

  if (!__BROWSER__ && options.isTS) {
    const { expressionPlugins } = resolvedOptions
    if (!expressionPlugins.includes('typescript')) {
      resolvedOptions.expressionPlugins = [...(expressionPlugins || []), 'typescript']
    }
  }

  const {
    body: [statement]
  } = isString(source)
    ? parse(source, {
        sourceType: 'module',
        plugins: resolvedOptions.expressionPlugins
      }).program
    : source
  let children!: JSXElement[] | JSXFragment['children']
  if (statement.type === 'ExpressionStatement') {
    children =
      statement.expression.type === 'JSXFragment'
        ? statement.expression.children
        : statement.expression.type === 'JSXElement'
          ? [statement.expression]
          : []
  }

  const ast: RootNode = {
    type: IRNodeTypes.ROOT,
    children,
    source: isString(source) ? source : '', // TODO
    components: [],
    directives: [],
    helpers: new Set(),
    temps: 0
  }

  const [nodeTransforms, directiveTransforms] = getBaseTransformPreset(prefixIdentifiers)

  const ir = transform(
    ast,
    extend({}, resolvedOptions, {
      nodeTransforms: [
        ...nodeTransforms,
        ...(options.nodeTransforms || []) // user transforms
      ],
      directiveTransforms: extend(
        {},
        directiveTransforms,
        options.directiveTransforms || {} // user transforms
      )
    })
  )

  return generate(
    ir as unknown as VaporRootIRNode,
    resolvedOptions
  ) as unknown as VaporCodegenResult
}

interface JsxCompilerOptions {
  /**
   * jsx parser
   * @param source - jsx code
   * @returns Babel AST
   */
  parser?: (source: string) => BabelProgram
}

export type CompilerOptions = HackOptions<BaseCompilerOptions> & JsxCompilerOptions
export type TransformPreset = [NodeTransform[], Record<string, DirectiveTransform>]

export function getBaseTransformPreset(_prefixIdentifiers?: boolean): TransformPreset {
  return [
    [
      transformText,
      transformElement,
      // transformSlot,
      transformChildren
    ],
    {
      bind: transformBind,
      on: transformOn
      // model: transformModel,
      // show: transformShow,
      // html: transformHtml,
    }
  ]
}
