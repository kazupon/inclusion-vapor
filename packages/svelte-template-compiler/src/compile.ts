// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `@vue/compiler-vapor`
// Author: Evan you (https://github.com/yyx990803) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/vuejs/core-vapor
// Code url: https://github.com/vuejs/core-vapor/blob/6608bb31973d35973428cae4fbd62026db068365/packages/compiler-vapor/src/compile.ts

import { ErrorCodes, createCompilerError, defaultOnError } from '@vue-vapor/compiler-dom'
import { generate } from '@vue-vapor/compiler-vapor'
import { extend, isString } from '@vue-vapor/shared'
import { IRNodeTypes, isSvelteParseError } from './ir/index.ts'
import { transform } from './transform.ts'
import {
  transformBind,
  transformChildren,
  transformComment,
  transformElement,
  transformFor,
  transformHtml,
  transformIf,
  transformModel,
  transformOn,
  transformSlot,
  transformSlotOutlet,
  transformTemplateRef,
  transformText
} from './transforms/index.ts'

import type { CompilerError, SourceLocation } from '@vue-vapor/compiler-dom'
import type {
  CompilerOptions as BaseCompilerOptions,
  VaporCodegenResult,
  RootIRNode as VaporRootIRNode
} from '@vue-vapor/compiler-vapor'
import type { RootNode, SvelteCompileError, SvelteTemplateNode } from './ir/index.ts'
import type { DirectiveTransform, HackOptions, NodeTransform } from './transforms/index.ts'

// Svelte Template Code / Svelte Template AST -> IR (transform) -> JS (generate)
export function compile(
  source: string | SvelteTemplateNode,
  options: CompilerOptions = {} // TODO: maybe we need some svelte compiler options
): VaporCodegenResult {
  const onError = options.onError || defaultOnError
  const isModuleMode = options.mode === 'module'
  /* istanbul ignore if */
  if (__BROWSER__) {
    if (options.prefixIdentifiers === true) {
      onError(createCompilerError(ErrorCodes.X_PREFIX_ID_NOT_SUPPORTED))
    } else if (isModuleMode) {
      onError(createCompilerError(ErrorCodes.X_MODULE_MODE_NOT_SUPPORTED))
    }
  }

  const prefixIdentifiers = !__BROWSER__ && (options.prefixIdentifiers === true || isModuleMode)

  if (options.scopeId && !isModuleMode) {
    onError(createCompilerError(ErrorCodes.X_SCOPE_ID_NOT_SUPPORTED))
  }

  const resolvedOptions = extend({}, options, {
    prefixIdentifiers
  })

  let svelteAst = {} as SvelteTemplateNode
  try {
    svelteAst = parse(source, resolvedOptions)
  } catch (error: unknown) {
    if (isSvelteParseError(error) && isString(source)) {
      const vaporError = error as unknown as CompilerError
      vaporError.loc = convertToVaporCompileErrorSourceLocation(source, error)
      onError(vaporError)
    } else {
      onError(error as CompilerError)
    }
  }

  const ast: RootNode = {
    type: IRNodeTypes.ROOT,
    children: svelteAst.children || [],
    source: isString(source) ? source : '', // TODO:
    components: [],
    directives: [],
    helpers: new Set(),
    temps: 0
  }

  const [nodeTransforms, directiveTransforms] = getBaseTransformPreset(prefixIdentifiers)

  const ir = transform(
    ast,
    extend({}, resolvedOptions, {
      nodeTransforms: [
        ...nodeTransforms,
        ...(options.nodeTransforms || []) // user transforms
      ],
      directiveTransforms: extend(
        {},
        directiveTransforms,
        options.directiveTransforms || {} // user transforms
      )
    })
  )

  return generate(ir as unknown as VaporRootIRNode, resolvedOptions)
}

function parse(source: string | SvelteTemplateNode, options: CompilerOptions): SvelteTemplateNode {
  return isString(source) ? getSvelteTemplateNode(source, options) : source
}

function convertToVaporCompileErrorSourceLocation(
  source: string,
  error: SvelteCompileError
): SourceLocation {
  return {
    start: {
      offset: error.start.character, // TODO: is this correct?
      line: error.start.line,
      column: error.start.column
    },
    end: {
      offset: error.end.character, // TODO: is this correct?
      line: error.end.line,
      column: error.end.column
    },
    source
  }
}

function getSvelteTemplateNode(
  source: string,
  { parser }: SvelteCompilerOptions
): SvelteTemplateNode {
  if (!parser) {
    throw new Error('svelte code parsing function option is not given.')
  }
  return parser(source)
}

interface SvelteCompilerOptions {
  /**
   * Svelte parser
   * @param source - Svelte code
   * @returns Svelte template AST
   */
  parser?: (source: string) => SvelteTemplateNode
}

export type CompilerOptions = HackOptions<BaseCompilerOptions> & SvelteCompilerOptions
export type TransformPreset = [NodeTransform[], Record<string, DirectiveTransform>]

export type { VaporCodegenResult } from '@vue-vapor/compiler-vapor'

export function getBaseTransformPreset(_prefixIdentifiers?: boolean): TransformPreset {
  return [
    [
      // transformOnce,
      transformIf,
      transformFor,
      transformSlotOutlet,
      transformTemplateRef,
      transformText,
      transformElement,
      transformSlot,
      transformComment,
      transformHtml,
      transformChildren
    ],
    {
      bind: transformBind,
      on: transformOn,
      // text: transformText,
      // show: transformShow,
      model: transformModel
    }
  ]
}
