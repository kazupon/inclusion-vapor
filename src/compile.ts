import { parse } from 'svelte/compiler'
import { ErrorCodes, createCompilerError, defaultOnError } from '@vue-vapor/compiler-dom'
import { generate } from '@vue-vapor/compiler-vapor'
import { extend, isString } from '@vue-vapor/shared'
import { transform } from './transform'
import { IRNodeTypes } from './ir'

import type {
  CompilerOptions,
  VaporCodegenResult,
  RootIRNode as VaporRootIRNode
} from '@vue-vapor/compiler-vapor'
import type { SvelteAst, RootNode } from './ir'
import type { NodeTransform, DirectiveTransform } from './transforms'

// Svelte Code / Svelte AST -> IR (transform) -> JS (generate)
export function compile(
  source: string | SvelteAst,
  options: CompilerOptions = {} // TODO: maybe we need some svelte compiler options
): VaporCodegenResult {
  const onError = options.onError || defaultOnError
  const isModuleMode = options.mode === 'module'
  /* istanbul ignore if */
  if (__BROWSER__) {
    if (options.prefixIdentifiers === true) {
      onError(createCompilerError(ErrorCodes.X_PREFIX_ID_NOT_SUPPORTED))
    } else if (isModuleMode) {
      onError(createCompilerError(ErrorCodes.X_MODULE_MODE_NOT_SUPPORTED))
    }
  }

  const prefixIdentifiers = !__BROWSER__ && (options.prefixIdentifiers === true || isModuleMode)

  if (options.scopeId && !isModuleMode) {
    onError(createCompilerError(ErrorCodes.X_SCOPE_ID_NOT_SUPPORTED))
  }

  const resolvedOptions = extend({}, options, {
    prefixIdentifiers
  })

  // TODO: tweak options for svelte parser
  const svelteAst = isString(source) ? parse(source) : source
  // TODO: convert `instance` and `module` to the Vapor (Vue) runtime API
  // svelteAst.instance
  // svelteAst.module

  const ast: RootNode = {
    type: IRNodeTypes.ROOT,
    children: svelteAst.html.children || [],
    source: isString(source) ? source : '', // TODO
    components: [],
    directives: [],
    helpers: new Set(),
    temps: 0
  }

  const [nodeTransforms, directiveTransforms] = getBaseTransformPreset(prefixIdentifiers)

  const ir = transform(
    ast,
    // @ts-expect-error -- FIXME: TS2345
    extend({}, resolvedOptions, {
      nodeTransforms: [
        ...nodeTransforms,
        ...(options.nodeTransforms || []) // user transforms
      ],
      directiveTransforms: extend(
        {},
        directiveTransforms,
        options.directiveTransforms || {} // user transforms
      )
    })
  )

  return generate(ir as unknown as VaporRootIRNode, resolvedOptions)
}

// export type CompilerOptions = HackOptions<BaseCompilerOptions>
export type TransformPreset = [NodeTransform[], Record<string, DirectiveTransform>]

export function getBaseTransformPreset(_prefixIdentifiers?: boolean): TransformPreset {
  return [
    [
      // transformVOnce,
      // transformVIf,
      // transformVFor,
      // transformSlotOutlet,
      // transformTemplateRef,
      // transformText,
      // transformElement,
      // transformVSlot,
      // transformComment,
      // transformChildren,
    ],
    {
      // bind: transformVBind,
      // on: transformVOn,
      // html: transformVHtml,
      // text: transformVText,
      // show: transformVShow,
      // model: transformVModel,
    }
  ]
}
