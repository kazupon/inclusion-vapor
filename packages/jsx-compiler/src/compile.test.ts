import { describe, test, expect } from 'vitest'
import { parse } from '@babel/parser'
import { compile } from './compile'

const jsxCode = `
(<button onClick={increment}>
  count is {count}
</button>)`

describe('compile', () => {
  test('source code', () => {
    const { code } = compile(jsxCode, {
      sourceMap: true,
      parser: (source: string) =>
        parse(source, {
          sourceType: 'module',
          plugins: ['jsx']
        }).program
    })
    expect(code).toMatchSnapshot('code')
    expect(code).contains(`_renderEffect(() => _setText(n0, "\\n  count is ", count, "\\n"))`)
  })

  test('ast', () => {
    const { code } = compile(
      parse(jsxCode, {
        sourceType: 'module',
        plugins: ['jsx']
      }).program,
      {
        sourceMap: true
      }
    )
    expect(code).toMatchSnapshot('code')
    expect(code).contains(`_renderEffect(() => _setText(n0, "\\n  count is ", count, "\\n"))`)
  })
})
