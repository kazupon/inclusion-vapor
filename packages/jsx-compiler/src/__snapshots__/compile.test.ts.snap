// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compile > ast > code 1`] = `
"import { delegate as _delegate, renderEffect as _renderEffect, setText as _setText, delegateEvents as _delegateEvents, template as _template } from 'vue/vapor';
const t0 = _template("<button></button>")
_delegateEvents("click")

export function render(_ctx) {
  const n0 = t0()
  _delegate(n0, "click", () => increment)
  _renderEffect(() => _setText(n0, "\\n  count is ", count, "\\n"))
  return n0
}"
`;

exports[`compile > source code > code 1`] = `
"import { delegate as _delegate, renderEffect as _renderEffect, setText as _setText, delegateEvents as _delegateEvents, template as _template } from 'vue/vapor';
const t0 = _template("<button></button>")
_delegateEvents("click")

export function render(_ctx) {
  const n0 = t0()
  _delegate(n0, "click", () => increment)
  _renderEffect(() => _setText(n0, "\\n  count is ", count, "\\n"))
  return n0
}"
`;
